### PR TITLE
webdriverio: workaround for Safari to moveTo, getLocation, dragAndDrop, getSize

### PR DESCRIPTION
**Pull request by mgrybyk**
## Proposed changes

Because of defect in safari https://github.com/SeleniumHQ/selenium/issues/6637 `getElementRect` is not return `y`.
We are calculating x and y offset so that is one of x, y, width, height is not defined will get `NaN` as a result.

Proposal: try get rect with javascript or set 0 as default value in case on of x, y, width, height is not defined

Affected browser: Safari 12
Affected commands: moveTo, getLocation, dragAndDrop, getSize

fix for #3608 

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/master/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by mgrybyk** [comment]
We can also "improve" getLocation behavior to work in hacky way so that it will try get location with javascript if regular getLocation / getRect failed to do it. But I would not like to do that


```
// ...
    let locationJs = { x: 0, y: 0, width: 0, height: 0 }

    // if one of [x, y, height, width] is undefined get location with javascript
    if (Object.keys(locationJs).some(key => location[key] == null)) {
        locationJs = await browser.execute(function (el) {
            const { left, top, width, height } = el.getBoundingClientRect();
            return {
                x: left + this.scrollX,
                y: top + this.scrollY,
                width,
                height
            };
        }, this)

        // try set proper value
        Object.keys(locationJs).forEach(key => {
            if (location[key] == null) {
                location[key] = typeof locationJs[key] === 'number' ? Math.floor(locationJs[key]) : 0
            }
        })
    }
```

**Comment by flunderpero** [comment]
This would help, next command I need is `dragAndDrop` which will likely not work either. Other commands might suffer from this bug, too.
In my humbled opinion it would be better to fix `getElementRect`. 

Tomorrow I will try to use the TechnologyPreview of Safari to see if the bug is gone with that.

**Comment by mgrybyk** [comment]
@flunderpero getLocation and dragAndDrop will not work

**Comment by mgrybyk** [comment]
I'm not sure if there is a value in this PR in such case.

**Comment by mgrybyk** [comment]
@flunderpero this should fix `getElementRect`.

However I'm not sure if it is a good idea to add such hacks to the framework

**Comment by abjerstedt** [comment]
I'm honestly not sure how i feel about this fix. I got blocked by this exact bug last week, but I don't think we should use an override just to support safari 12 (until the next release). 

I think the better behavior is to put a hack into any dependency commands to have a default value of 0 in the event that the driver returns a null.

**Comment by mgrybyk** [comment]
I think both solutions are not really good. 
Return 0 instead of undefined will not work either.

It would be great to throw exception instead because we are expecting to get some value and failed.

Closing this PR for now

**Comment by christian-bromann** [comment]
> Return 0 instead of undefined will not work either.

Why is that?

**Comment by mgrybyk** [comment]
> Why is that?

imagine there is a button in the middle of the screen with coords x: 600, y: 600.

moveTo (and other commands that use getRect) won't work because x: 600, y: 0 are not valid coords of the element.

**Comment by christian-bromann** [comment]
> moveTo (and other commands that use getRect) won't work because x: 600, y: 0 are not valid coords of the element.

ok, I was assuming that Safaridriver only fails returning `y` in certain edge cases not all the time.

**Comment by abjerstedt** [comment]
@christian-bromann Yeahhh.... Safaridriver for v12 is just flat omitting the Y coordinate in getElementRect. 

I logged a bug directly with apple, and they asked for stacktraces :-/. However, it sounds like they already fixed it for the next safari release.

**Comment by christian-bromann** [comment]
I think it makes sense to include this workaround for all safari sessions until majority of people use the fixed safaridriver, what do you think @mgrybyk ?

**Comment by mgrybyk** [comment]
@christian-bromann I don't mind, it is up to you to decide.
For me fix looks more or less clean and we can always replace function `getElementRect` with `scope.getElementRect(scope.elementId)`

**Comment by mgrybyk** [comment]
And even better, once it is fixed in Safari we can leave verification of getElementRect to be sure it has returned x, y, height and width

**Comment by christian-bromann** [comment]
@mgrybyk Let's reopen then

**Comment by christian-bromann** [comment]
@mgrybyk @abjerstedt any objections to merge this?

File: packages/webdriverio/tests/commands/element/moveTo.test.js [hunk 1]
```diff
@@ -35,6 +35,21 @@ describe('moveTo', () => {
...
+    it('should do a moveTo with params if getElementRect returned empty object', async () => {
+        const browser = await remote({
+            baseUrl: 'http://foobar.com',
+            capabilities: {
+                browserName: 'foobar'
+            }
+        })
+
+        const elem = await browser.$('#elem')
+        request.setMockResponse([{}, { x: 5, y: 10, height: 33, width: 44 }])
```
Comment (christian-bromann) [inline packages/webdriverio/tests/commands/element/moveTo.test.js]: How is this testing the edge case behavior if all values are given? Shouldn't one be omitted?

Comment (mgrybyk) [reply packages/webdriverio/tests/commands/element/moveTo.test.js]: this is the case when getElementRect has not returned any value but execute script returned all the values.

@christian-bromann btw, what should we do in case we have failed to get one of x,y,width or height with JS. Currently it is fallback to 0. But I think we'd better throw an exception.

Can you judge this? :)
@abjerstedt @christian-bromann

Comment (christian-bromann) [reply packages/webdriverio/tests/commands/element/moveTo.test.js]: I think `el.getBoundingClientRect` will always return proper values therefor I would also say we can remove `if (!el || !el.getBoundingClientRect) return`, no?

Comment (mgrybyk) [reply packages/webdriverio/tests/commands/element/moveTo.test.js]: @christian-bromann I have a failing test because of that
`TypeError: el.getBoundingClientRect is not a function`

don't know how to mock that. Anyway it is DOM interaction

Comment (christian-bromann) [reply packages/webdriverio/tests/commands/element/moveTo.test.js]: so if the element isn't found we should throw or are you talking about webdriverio unit test. Also could you rebase the branch to current master?

Comment (mgrybyk) [reply packages/webdriverio/tests/commands/element/moveTo.test.js]: I'm talking about both.

First of all we should throw an exception if getElementRect failed and then getBoundingClientRect failed as well (for whatever reason)

Secondly I have problems with unit test, but I think additional verification is not a problem `if (!el || !el.getBoundingClientRect) return`

Doing rebase, added test for throw error, and will push after it soon

Comment (mgrybyk) [reply packages/webdriverio/tests/commands/element/moveTo.test.js]: @christian-bromann updated

File: packages/webdriverio/src/utils.js [hunk 2]
```diff
@@ -393,3 +395,43 @@ export function verifyArgsAndStripIfElement(args) {
...
+        }, scope)
+
+        // try set proper value
+        Object.keys(defaults).forEach(key => {
+            if (rect[key] == null && typeof rectJs[key] === 'number') {
+                rect[key] = Math.floor(rectJs[key])
+            } else {
+                log.error('getElementRect', rect)
+                log.error('getBoundingClientRect', rectJs)
+                throw new Error('Failed to getElementRect')
```
Comment (christian-bromann) [inline packages/webdriverio/src/utils.js]:
```suggestion
                throw new Error('Failed to receive element rects via execute command')
```

File: packages/webdriverio/src/utils.js [hunk 3]
```diff
@@ -393,3 +395,43 @@ export function verifyArgsAndStripIfElement(args) {
...
+                x: left + this.scrollX,
+                y: top + this.scrollY,
+                width,
+                height
+            }
+        }, scope)
+
+        // try set proper value
+        Object.keys(defaults).forEach(key => {
+            if (rect[key] == null && typeof rectJs[key] === 'number') {
```
Comment (christian-bromann) [inline packages/webdriverio/src/utils.js]:
```suggestion
            if (rect[key] === null && typeof rectJs[key] === 'number') {
```

File: packages/webdriverio/src/utils.js [hunk 4]
```diff
@@ -393,3 +395,43 @@ export function verifyArgsAndStripIfElement(args) {
...
+export async function getElementRect(scope) {
+    const rect = await scope.getElementRect(scope.elementId)
+
+    let defaults = { x: 0, y: 0, width: 0, height: 0 }
+
+    /**
+     * getElementRect workaround for Safari 12.0.3
+     * if one of [x, y, height, width] is undefined get rect with javascript
+     */
+    if (Object.keys(defaults).some(key => rect[key] == null)) {
```
[the remark this answers is not in the document]
Comment (mgrybyk) [reply packages/webdriverio/src/utils.js]: @christian-bromann I want to perform check for null and undefined.
I can do it easily in one line rect[key] == null
or can do rect[key] === null and rect[key] === undefined

I'd better keep rect[key] == null, if you don't mind

File: packages/webdriverio/src/utils.js (hunk 3 above)
Comment (mgrybyk) [reply packages/webdriverio/src/utils.js]: @christian-bromann same here

**Comment by zuhito** [comment]
The fixed version of this problem is v5.x.x only?
I'd like to use the fixed version in v4.x.x. 🙏

**Comment by christian-bromann** [comment]
> I'd like to use the fixed version in v4.x.x. 🙏

Please provide a PR and I can release it for v4.

**Comment by zuhito** [comment]
I tried to change the code for v4.x.x but I seem not to be able to use getBoundingClientRect() in the code...
